### PR TITLE
[wip] [ci] Run tests in a full-fledged ubuntu 19.10 VM

### PR DESCRIPTION
**Pull request by njsmith**
I'm not sure if this is actually useful... I just got curious about it
today and it came together more easily than I expected, so now I'm
curious if it actually works :-).

Rumor says that Travis supports nested KVM acceleration now, if you're
using their 'bionic' image, so in theory this should work and be
reasonably fast?

Right now, it isn't especially useful. But if we started adding
support for new Linux kernel features like pidfd (#1241) or
io_uring (#932), then this would become important, because AFAICT this
is the only way to run tests on new kernels on any of the hosted CI
services.

I guess we could also use it to like, test on ARM or something, if we
cared? I experimented with testing on 32-bit, but the test setup
failed because cryptography has stopped shipping 32-bit wheels, and I
couldn't be bothered to figure out how to fix it.

### CODE REVIEW

**Comment by njsmith** [comment]
Since the Trio run loop is essentially a kernel for a specialized python machine, I guess this is testing a virtual machine, running inside a virtual machine, running inside a virtual machine, running inside a real machine.

**Comment by pquentin** [comment]
Just trying to understand: does that mean we were not testing the latest Ubuntu in the py37_latest_ubuntu Azure Pipelines job? I'm asking because the new setup seems to be more complicated

**Comment by njsmith** [comment]
Because of how docker works, we were testing the latest Ubuntu userspace, but still using the CI system's kernel (which for Azure Pipelines I think is either from Ubuntu 16.04 or Ubuntu 18.04). For 99% of things that's fine, but if you want to test new kernel features, then you need a new kernel. This approach tests the latest Ubuntu userspace *running on the latest Ubuntu kernel*, so it does everything the docker approach does, and a bit more.

It's definitely more complicated and slower, though, so it depends on how much we care about testing on newer kernels. If we were to say, add a whole new `io_uring` backend, then we would care a lot :-). Right now though it's not so obvious.

**Comment by pquentin** [comment]
Ah, yes, that makes sense! Okay, so we have a complicated thing that took some time to get right, but now works and might be useful in the feature. Per the sunken cost fallacy, I think we should merge this!

But first, are we sure that our codecov setup still works? I still see references to `CODECOV_NAME` in ci.sh.

**Comment by auvipy** [comment]
does ubuntu LXD containers are of any use for it? https://github.com/fnordahl/travis-lxd , https://docs.travis-ci.com/user/reference/overview/ , https://fnordahl.com/2017/04/08/continuous-deployment-testing-of-github-hosted-code/

**Comment by njsmith** [comment]
@pquentin 
> Ah, yes, that makes sense! Okay, so we have a complicated thing that took some time to get right, but now works and might be useful in the feature. Per the sunk cost fallacy, I think we should merge this!

Heh :-). I was also thinking maybe we should close it, and then in the future if we have a need for it we'll know it's here and can resurrect the branch.

> But first, are we sure that our codecov setup still works? I still see references to CODECOV_NAME in ci.sh.

Whoops, good catch, thanks!

@auvipy 
> does ubuntu LXD containers are of any use for it?

I believe that LXD containers use the same basic technology as docker, i.e., the containers all share the same underlying host kernel.

File: .travis.yml [hunk 1]
```diff
@@ -11,9 +11,23 @@ matrix:
...
+    # kernel, which is important when we use new kernel features. This
+    # is also good for testing the latest openssl etc., and getting
+    # early warning of any issues that might happen in the next Ubuntu
+    # LTS.
+    - language: generic
+      # We use bionic for the host, b/c rumor says that Travis's
+      # 'bionic' systems have nested KVM enabled.
+      dist: bionic
+      env:
+        - "JOB_NAME=Ubuntu 19.10, full VM"
```
Comment (pquentin) [inline .travis.yml]: codecov does not like spaces here, and only shows "Ubuntu" - see https://codecov.io/gh/python-trio/trio/commit/7cabd162b3fe04bc1465cecca0a7b07804567210/build

Comment (pquentin) [reply .travis.yml]: However it can't be the exact issue as we have the spaces for the Azure Pipeliens jobs.

Comment (njsmith) [reply .travis.yml]: Hmm, it looks like something is going wrong at the Travis level... at the very top of the file, it shows the initial environment for the job as created by travis, and [that says](https://travis-ci.org/python-trio/trio/jobs/608056847#L216):

```
JOB_NAME=Ubuntu
```

Maybe I need to add more quotes to the YAML or something?

Comment (njsmith) [reply .travis.yml]: Yeah, adding more quotes seems to have fixed it.

Comment (pquentin) [reply .travis.yml]: Nice!

**Comment by pquentin** [comment]
Thanks, merged!

**Pull request merged**